KEYCLOAK-14103 Document setting of SameSite to None for Wildfly/EAP a…

mhajas opened the pull request:
…nd Tomcat

mhajas committed:
KEYCLOAK-14103 Document setting of SameSite to None for Wildfly/EAP and Tomcat

hmlnarik commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+[[_saml-jboss-adapter-samesite-setting]]
+===== Setting SameSite value for JSESSIONID cookie
+
+SAML Post binding might be broken in the latest versions of browsers due to changes to the SameSite default value. To
```
Explain a bit what SameSite is before using it.

hmlnarik commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+
+ samesite-cookie(mode=None, cookie-pattern=JSESSIONID)
+
+The support for this configuration is available in Wildfly from version 19.1.0 and in EAP from versions 7.3.2.GA and
```
Neither EAP 7.3.2 nor 7.2.9 is available. Refer the reader to EAP documentation for determining when this can be used, or don't reference EAP at all.

hmlnarik commented on securing_apps/topics/saml/java/tomcat-adapter/tomcat-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,21 @@
+[[_saml-tomcat-adapter-samesite-setting]]
+===== Setting SameSite value for JSESSIONID cookie
+
+SAML Post binding might be broken in the latest versions of browsers due to changes to the SameSite default value. To
```
Explain a bit what SameSite is before using it.

vmuzikar reviewed (commented):
I believe we should add instructions for Jetty too. If we decide to not to do this in this PR, we'll need to create a follow-up JIRA.

vmuzikar reviewed (commented):
Two more things.

Should we advice the users to strongly consider the potential security risk that `SameSite=None` brings (especially in Tomcat's case where this setting is global for all cookies)? Like some link to where SameSite is documented with a hint to use OIDC if they don't want to a security issue with SameSite?

Should we add some note that users don't actually need to upgrade their adapter to a newer version just to resolve the SameSite issue? I mean, all we did at the adapter's side is we adde a warning into the log.

andymunro commented on release_notes/topics/11_0_0.adoc:
```diff
@@ -0,0 +1,9 @@
+== SAML Post binding is broken in the latest versions of browsers
+
+The `SameSite` value `None` for `JSESSIONID` cookie is necessary for correct behavior of the Keycloak SAML adapter.
+Usage of a different value is causing resetting of the container's session with each request to the Keycloak, when
```
remove "the" before "Keycloak"
Change "Keycloak" to "{project_name}"

andymunro commented on release_notes/topics/11_0_0.adoc:
```diff
@@ -0,0 +1,9 @@
+
+The `SameSite` value `None` for `JSESSIONID` cookie is necessary for correct behavior of the Keycloak SAML adapter.
+Usage of a different value is causing resetting of the container's session with each request to the Keycloak, when
+the SAML Post binging is used. Refer to following steps for
```
add "the" before "following steps"

andymunro commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+[[_saml-jboss-adapter-samesite-setting]]
+===== Setting SameSite value for JSESSIONID cookie
+
+Browsers are planning to set default value for the `SameSite` attribute for cookies to `Lax`. This means that
```
Add "the" before "the"
Try to always follow "this" with a noun, so it is clear what "this" refers to.   "This setting"

andymunro commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+===== Setting SameSite value for JSESSIONID cookie
+
+Browsers are planning to set default value for the `SameSite` attribute for cookies to `Lax`. This means that
+cookies will be sent to applications only if the request originates in the same domain. This behavior can affect SAML
```
Add "the" before "SAML"

andymunro commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+
+Browsers are planning to set default value for the `SameSite` attribute for cookies to `Lax`. This means that
+cookies will be sent to applications only if the request originates in the same domain. This behavior can affect SAML
+Post binding which may become non-functional. To preserve full functionality of the SAML adapter, it is recommended to
```
use "we recommend" instead of "it is recommended"  -- it is more active

andymunro commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+Browsers are planning to set default value for the `SameSite` attribute for cookies to `Lax`. This means that
+cookies will be sent to applications only if the request originates in the same domain. This behavior can affect SAML
+Post binding which may become non-functional. To preserve full functionality of the SAML adapter, it is recommended to
+set the `SameSite` value to `None` for `JSESSIONID` cookie created by your container. Not doing so may result in
```
add "the" before `JSESSIONID`

andymunro commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+cookies will be sent to applications only if the request originates in the same domain. This behavior can affect SAML
+Post binding which may become non-functional. To preserve full functionality of the SAML adapter, it is recommended to
+set the `SameSite` value to `None` for `JSESSIONID` cookie created by your container. Not doing so may result in
+resetting container's session with each request to {project_name}.
```
add "the" before "container's"

andymunro commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+set the `SameSite` value to `None` for `JSESSIONID` cookie created by your container. Not doing so may result in
+resetting container's session with each request to {project_name}.
+
+To set the `SameSite` value to `None` for `JSESSIONID` cookie in Wildfly/EAP add a file `undertow-handlers.conf`, with
```
add "the" before JSESSIONID
add a command after EAP

andymunro commented on securing_apps/topics/saml/java/jboss-adapter/jboss-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,17 @@
+resetting container's session with each request to {project_name}.
+
+To set the `SameSite` value to `None` for `JSESSIONID` cookie in Wildfly/EAP add a file `undertow-handlers.conf`, with
+following content, to the `WEB-INF` directory of your application.
```
add  "the" before "following"

andymunro commented on securing_apps/topics/saml/java/tomcat-adapter/tomcat-adapter-samesite-setting.adoc:
```diff
@@ -0,0 +1,21 @@
+[[_saml-tomcat-adapter-samesite-setting]]
+===== Setting SameSite value for JSESSIONID cookie
+
+Browsers are planning to set default value for the `SameSite` attribute for cookies to `Lax`. This means that
```
Please make the same changes as above.

andymunro reviewed (changes requested):
Looks good. Just a few changes suggested.

mhajas commented:
Thank you for review @vmuzikar!

> I believe we should add instructions for Jetty too. If we decide to not to do this in this PR, we'll need to create a follow-up JIRA.

I created a follow-up Jira: https://issues.redhat.com/browse/KEYCLOAK-14372

> Two more things.
> 
> Should we advice the users to strongly consider the potential security risk that `SameSite=None` brings (especially in Tomcat's case where this setting is global for all cookies)? Like some link to where SameSite is documented with a hint to use OIDC if they don't want to a security issue with SameSite?
>
> Should we add some note that users don't actually need to upgrade their adapter to a newer version just to resolve the SameSite issue? I mean, all we did at the adapter's side is we adde a warning into the log.

I added both. Let me know if this is what you meant.

mhajas commented:
@andymunro Thank you very much for the review. I incorporated all suggestions (I also added some more things based on the review from @vmuzikar). Could you please take a look again?

vmuzikar commented:
@mhajas Thanks for the update. Found one last nitpick. "POST" in "Post binding" should be written as capitals (POST binding). Looking at the rest of the guide, we always refer to it as "POST binding", so I think we should be consistent.

mhajas commented:
@vmuzikar Thanks, done

vmuzikar commented on release_notes/topics/11_0_0.adoc:
```diff
@@ -0,0 +1,10 @@
+= Highlights
+
+== SAML Post binding is broken in the latest versions of browsers
```
One last "Post" ;)
```suggestion
== SAML POST binding is broken in the latest versions of browsers
```

vmuzikar commented:
@hmlnarik @mhajas The last thing I'm missing is mentioning the SameSite support in EAP more explicitly, ideally incl. versions. But I understand we cannot mention any version before release. So should we create a follow-up JIRA?

vmuzikar reviewed (dismissed):
Created [a follow-up JIRA](https://issues.redhat.com/browse/KEYCLOAK-14746) for the EAP versions.

LGTM, thanks @mhajas!
